File: Compression with Python/BZip2Compression.py
import bz2
import time
import os
import ConvertTo

min_compression_level = 1
max_compression_level = 9
extension = ".bz2"

def compress(file_path: str, data_unit: ConvertTo.DataUnits):
    """Compress using the BZip2 algorithm."""

    log = open("log_compress.txt", "a")
    f_input = open(file_path, 'rb')

    for i in range(min_compression_level, max_compression_level + 1):
        start_timer = time.time()

        f_output_path = file_path + str(i) + extension

        f_output = bz2.open(f_output_path, 'wb', compresslevel=i)
        f_output.write(f_input.read())

        f_input.seek(0)    
        f_output.close()

        log.write(str(i) + " " + str(ConvertTo.byte_to(os.path.getsize(f_output_path), data_unit)) + " " + str("{0:.2f}".format(time.time() - start_timer)) + " BZIP2\n")

    f_input.close()
    log.close()

def decompress(file_path: str):
    """Decompress using the BZip2 algorithm."""

    log = open("log_decompress.txt", "a")
    f_input = bz2.open(file_path + str(max_compression_level) + extension, 'rb', compresslevel=max_compression_level)
    start_timer = time.time()

    f_output = open(file_path, 'wb')
    f_output.write(f_input.read())

    log.write(file_path + " " + str("{0:.2f}".format(time.time() - start_timer)) + " BZIP2\n")

    f_output.close()
    f_input.close()
    log.close()

File: Compression with C#/Compression/BZip2Compression.cs
using System.IO;
using System.Diagnostics;
using ICSharpCode.SharpZipLib.BZip2;

namespace Compression
{
    /// <summary>
    /// BZip2 compression algorithm.
    /// </summary>
    class BZip2Compression : CompressionAlgorithm
    {
        public BZip2Compression(string extension, int minCompressionLevel = 1, int maxCompressionLevel = 9)
            : base(extension, minCompressionLevel, maxCompressionLevel)
        {
        }

        /// <summary>
        /// Compresses a given file using the BZip2 Algorithm.
        /// </summary>
        /// <param name="fInput">The file to compress.</param>
        /// <param name="dataUnit">The unit that will be saved in the log file</param>
        public override void Compress(FileInfo fInput = null, DataUnits dataUnit = DataUnits.Byte)
        {
            StreamWriter fLog = File.AppendText("bzip2" + fInput.Name + ".txt");
            Stopwatch timer = new Stopwatch();
            FileInfo info = null;

            for(int i = MinCompressionLevel ; i <= MaxCompressionLevel ; i++)
            {
                timer.Start();

                using(FileStream fInputStream = fInput.OpenRead())
                    using(FileStream fOutputStream = File.Create(fInput.FullName + i.ToString() + Extension))
                        BZip2.Compress(fInputStream, fOutputStream, true, level:i);

                timer.Stop();

                info = new FileInfo(fInput.FullName + i + Extension);

                fLog.WriteLine(i + " " + Byte.ConvertTo(info.Length, dataUnit) + " " + timer.Elapsed.TotalSeconds.ToString("n2") + " CSharp");
            }
            fLog.Close();
        }

        /// <summary>
        /// Decompresses the given files using the BZip2 algorithm.
        /// </summary>
        /// <param name="fCompressed">The name of the file to decompress.</param>
        public override void Decompress(string fName = null)
        {
            FileInfo fInput = new FileInfo(fName + MaxCompressionLevel + Extension);
            StreamWriter fLog = File.AppendText("log_decompress.txt");
            Stopwatch timer = new Stopwatch();

            timer.Start();

            using(FileStream fInputStream = fInput.OpenRead())
                using(FileStream fOutputStream = File.Create(fName))
                    BZip2.Decompress(fInputStream, fOutputStream, true);

            timer.Stop();

            fLog.WriteLine(fName + timer.Elapsed.TotalSeconds.ToString("n2") + " BZIP2");

            fLog.Close();
        }
    }
}
